flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/async/CompletedOperationCache.java

```diff
@@ -144,21 +146,35 @@ private boolean isRunning() {
     }
 
     /**
-     * Returns the operation result or a {@code Throwable} if the {@code CompletableFuture}
-     * finished, otherwise {@code null}.
+     * Returns a future holding the result of the operation with the specified key. It may already
+     * be complete, and may complete exceptionally with a {@code Throwable} in case the operation
+     * fails.
      *
      * @throws UnknownOperationKeyException If the operation is not found, and there is no ongoing
      *     operation under the provided key.
      */
-    @Nullable
-    public Either<Throwable, R> get(final K operationKey) throws UnknownOperationKeyException {
+    public CompletableFuture<R> get(final K operationKey) throws UnknownOperationKeyException {
         ResultAccessTracker<R> resultAccessTracker;
-        if ((resultAccessTracker = registeredOperationTriggers.get(operationKey)) == null
-                && (resultAccessTracker = completedOperations.getIfPresent(operationKey)) == null) {
-            throw new UnknownOperationKeyException(operationKey);
+
+        if ((resultAccessTracker = completedOperations.getIfPresent(operationKey)) != null) {
+            Either<Throwable, R> result = resultAccessTracker.accessOperationResultOrError();
+
+            if (result == null) {
+                return CompletableFuture.failedFuture(
```

Comment:
Not quite true: If the operation is not finished (i.e. is still in `registeredOperationTriggers`), then the unfinished future is returned.

I think this case will disappear when refactoring the `ResultAccessTracker` to just use the future instead of this "result or error" construct